Eq params

pblivingston opened the pull request:
Closes #14 

## What

Adds Eq class as base for BusEq and StripEq with:
- on
- ab
- channel[i].cell[j].on
- channel[i].cell[j].type
- channel[i].cell[j].f
- channel[i].cell[j].gain
- channel[i].cell[j].q
- Load($filename)
- Save($filename)

## Test summary

Pester tests pass for all kinds, no manual tests needed

## Questions for reviewers

For consistency with Vban and Eq, I plan to submit a small PR moving 'identifier' to device.ps1 or I could do that here if you think it's relevant enough. If you'd rather Eq and Device have their identifier methods in bus.ps1 and strip.ps1, I can make the inverse change here instead.

pblivingston committed:
Create eq.ps1

- eq
- eq channel
- eq cell

pblivingston committed:
stripeq, buseq

- replace iremote with eq for stripeq and buseq
- move identifier to eq
- avoid passing entire parent objects

pblivingston committed:
update tests

pester tests pass for all kinds

pblivingston committed:
update docs

onyx-and-iris commented:
Hi, regarding how we structure the identifiers, I think we can do either. I like your implementation but I think there are two potential benefits to keeping the identifiers in strip.ps1 and bus.ps1:

- Since each class identifier is the full route to the API command I suspect it's more intuitive for someone else (unfamiliar with the codebase) reading the code exactly what that class is doing, without having to search parent classes for identifiers.
- The more you have to pass to an objects constructor the more classes must know about each other and the tighter the coupling.

I don't think either of those are a big deal here though, so we can go with your way if you like.

---

I do have a couple of small suggestions though. If we update the KindMap like so:

```powershell
$KindMap = @{
    'basic'  = @{
        ...
        'eq_ch'     = @{'strip' = 0; 'bus' = 0 }
        'cells'     = 0
    };
    'banana' = @{
        ...
        'eq_ch'     = @{'strip' = 0; 'bus' = 8 }
        'cells'     = 6
    };
    'potato' = @{
        ...
        'eq_ch'     = @{'strip' = 2; 'bus' = 8 }
        'cells'     = 6
    };
}
```

Then we can fetch the channel count from the kind maps by name:
```powershell
    Eq ([int]$index, [Object]$remote, [string]$kindOfEq) : base ($index, $remote) {
        $this.kindOfEq = $kindOfEq

        AddBoolMembers -PARAMS @('on', 'ab')

        $this.channel = @()
        for ($ch = 0; $ch -lt $remote.kind.eq_ch[$this.kindOfEq]; $ch++) {
            $this.channel.Add([EqChannel]::new($ch, $remote, $this.identifier()))
        }
    }
```

I also suggest renaming $prefix to $kindOfEq. This reduces the amount the Eq constructor must know and takes advantage of the fact we're already passing the kind of eq.

Let me know your thoughts.

onyx-and-iris commented:
If we keep this implementation for the identifiers which I'm happy to go with, then please submit the change to device class in a separate PR.

pblivingston committed:
identifier, kindOfEq

- move identifier back to BusEq and StripEq for clarity and looser coupling
- adjust kindmap so we can get channel count with kindOfEq

pblivingston commented:
Good points about clarity and looser coupling. Fetching channel counts by name is nice as well. Changes implemented in latest commit.

pblivingston committed:
forgot to save kinds

onyx-and-iris commented:
Great PR thanks! Good idea to include EQ saving/loading tests.
